src/java.base/share/classes/javax/crypto/spec/HPKEParameterSpec.java

```diff
@@ -0,0 +1,480 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package javax.crypto.spec;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.Key;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Objects;
+
+/**
+ * This immutable class specifies the set of parameters used with a {@code Cipher} for the
+ * <a href="https://www.rfc-editor.org/info/rfc9180">Hybrid Public Key Encryption</a>
+ * (HPKE) algorithm. The <a href=
+ * "{@docRoot}/../specs/security/standard-names.html#cipher-algorithms">
+ * standard algorithm name</a> for the cipher is "HPKE".
+ * <p>
+ * In HPKE, the sender's {@code Cipher} is always initialized with the
+ * recipient's public key in {@linkplain Cipher#ENCRYPT_MODE encrypt mode},
+ * while the recipient's {@code Cipher} object is initialized with its own
+ * private key in {@linkplain Cipher#DECRYPT_MODE decrypt mode}.
+ * <p>
+ * An {@code HPKEParameterSpec} object can be provided at HPKE
+ * {@linkplain Cipher#init(int, Key, AlgorithmParameterSpec) cipher initialization}.
+ * <p>
+ * An {@code HPKEParameterSpec} object can be created in two ways:
+ * <ul>
+ * <li> {@link #of()} creates an instance with unspecified KEM, KDF, and AEAD
+ * algorithms, which will be determined by the implementation based on the key
+ * provided to {@code init()}.
+ * <li> {@link #of(int, int)} creates an instance with explicitly specified
+ * KDF, and AEAD algorithm identifiers, but the KEM algorithm identifier will be
+ * determined by the implementation based on the key provided to {@code init()}.
+ * <li> {@link #of(int, int, int)} creates an instance with explicitly
+ * specified KEM, KDF, and AEAD algorithm identifiers.
+ * </ul>
+ * The terms "KEM algorithm identifiers", "KDF algorithm identifiers", and
+ * "AEAD algorithm identifiers" refer to their respective numeric values
+ * (specifically, {@code kem_id}, {@code kdf_id}, and {@code aead_id}) as
+ * defined in <a href="https://www.rfc-editor.org/rfc/rfc9180.html#section-7">Section 7</a>
+ * of RFC 9180 and the
+ * <a href="https://www.iana.org/assignments/hpke/hpke.xhtml">IANA HPKE page</a>.
+ * <p>
+ * Once an {@code HPKEParameterSpec} object is created, additional methods
```

Comment:
Are these additional methods commonly called? If so, did you consider using something like a Builder to avoid the extra object creations?

Reply:
`info` and `encapsulation` are frequently called. I don't think it's worth using a builder. The final `build()` call is always the same and unnecessary.

Reply:
If they are common, would it make more sense to include them as parameters to an overloaded version of the static `of` method?

Reply:
I'd rather not. There are already three `of` methods. Adding `info` would make make six, and adding `encapsulation` would be twelve. I cannot say which combination is more popular than another one.

Also,  the arguments of the current `of` methods are all about algorithm identifiers. Adding `info` and `encapsulation` to the argument list could make the API harder to use, as the method signature only includes the types of the parameters but not their meanings. While providing them through the `info` and `encapsulation` methods are very clean. They also have accompanying "getter" methods with the same names.